lib/ab_test.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+class AbTest
+  attr_reader :buckets, :experiment_name, :default_bucket, :should_log
+
+  MAX_SHA = (16 ** 64) - 1
+
+  # @param [Proc<String>,RegExp,string,Boolean,nil] should_log Controls whether bucket data for this
+  #                                                            A/B test is logged with specific
+  #                                                            events.
+  # @yieldparam [ActionDispatch::Request] request
+  # @yieldparam [String,nil] service_provider Issuer string for the service provider associated with
+  #                                           the current session.
+  # @yieldparam [User] user
+  # @yieldparam [Hash] user_session
+  def initialize(
+    experiment_name:,
+    buckets: {},
+    should_log: nil,
+    default_bucket: :default,
+    &discriminator
+  )
+    @buckets = buckets
+    @discriminator = discriminator
+    @experiment_name = experiment_name
+    @default_bucket = default_bucket
+    @should_log = should_log
+    raise 'invalid bucket data structure' unless valid_bucket_data_structure?
+    ensure_numeric_percentages
+    raise 'bucket percentages exceed 100' unless within_100_percent?
+  end
+
+  # @param [ActionDispatch::Request] request
+  # @param [String,nil] service_provider Issuer string for the service provider associated with
+  #                                      the current session.
+  # @params [Hash] session
+  # @param [User] user
+  # @param [Hash] user_session
+  def bucket(request:, service_provider:, session:, user:, user_session:)
+    return nil if no_percentages?
+
+    discriminator = @discriminator.call(request:, service_provider:, session:, user:, user_session:)
+
+    return nil if discriminator.blank?
+
+    user_value = percent(discriminator)
+
+    min = 0
+    buckets.keys.each do |key|
+      max = min + buckets[key]
+      return key if user_value > min && user_value <= max
+      min = max
+    end
+
+    @default_bucket
+  end
+
+  def include_in_analytics_event?(event_name)
+    if should_log.is_a?(Proc)
+      should_log.call(event_name)
+    elsif should_log.is_a?(Regexp)
+      should_log.match?(event_name)
+    elsif should_log.is_a?(String)
+      event_name == should_log
+    elsif should_log == true || should_log == false
+      should_log
+    elsif !should_log.nil?
+      raise 'Unexpected value used for should_log'
+    else
+      true
+    end
+  end
+
+  private
+
+  def no_percentages?
+    buckets.empty? || buckets.values.all? { |pct| pct == 0 }
+  end
+
+  def percent(discriminator)
+    Digest::SHA256.hexdigest("#{discriminator}:#{experiment_name}").to_i(16).to_f / MAX_SHA * 100
```

Comment:
so as long as we're here, an idea for the future. Doesn't have to be this PR but something on my mind

I think it would be nice for future, new AB tests to not use SHA256 because it's slow, we could find some other faster stable digest and use that. ([maybe something like this](https://github.com/Cyan4973/xxHash))

To maintain legacy compatibility and not redo existing AB tests, we would need to add a `use_legacy_digest:` that defaults to `false` and opt in our existing tests

Reply:
Yeah, I like this idea. Captured as [LG-14187](https://cm-jira.usa.gov/browse/LG-14187)